edlib/src/edlib.tpp

```diff
@@ -65,42 +69,66 @@ struct Block {
 
 
 /**
- * Defines equality relation on alphabet characters.
- * By default each character is always equal only to itself, but you can also provide additional equalities.
+ * Defines equality relation on indexes of alphabet elements.
+ * By default each alphabet index is always equal only to itself, but you can also provide additional equalities.
  */
 template<class AlphabetIdx>
 class EqualityDefinition {
-public:/*
-    EqualityDefinition(const string& alphabet,
-                       const EdlibEqualityPair* additionalEqualities = NULL,
-                       const int additionalEqualitiesLength = 0) {
-        for (int i = 0; i < static_cast<int>(alphabet.size()); i++) {
-            for (int j = 0; j < static_cast<int>(alphabet.size()); j++) {
-                matrix[i][j] = (i == j);
-            }
-        }
-        if (additionalEqualities != NULL) {
-            for (int i = 0; i < additionalEqualitiesLength; i++) {
-                size_t firstTransformed = alphabet.find(additionalEqualities[i].first);
-                size_t secondTransformed = alphabet.find(additionalEqualities[i].second);
-                if (firstTransformed != string::npos && secondTransformed != string::npos) {
-                    matrix[firstTransformed][secondTransformed] = matrix[secondTransformed][firstTransformed] = true;
-                }
-            }
-        }
-    }*/
-
-    /**
-     * @param a  Element from transformed sequence.
-     * @param b  Element from transformed sequence.
+private:
+    SetOfPairs<AlphabetIdx> equalitySet;
+    std::vector<bool> hasAdditionalEqualities;
+public:
+    EqualityDefinition(const SetOfPairs<AlphabetIdx>& equalitySet_, const std::vector<bool>& hasAdditionalEqualities_){
+        equalitySet = equalitySet_;
+        hasAdditionalEqualities = hasAdditionalEqualities_;
+    }
+    /*
+     * @param a  an alphabet index from transformed sequence.
+     * @param b  an alphabet index from transformed sequence.
      * @return True if a and b are defined as equal, false otherwise.
      */
-    // TODO: We need to implement areEqual later (to support additional equalities)
     bool areEqual(AlphabetIdx a, AlphabetIdx b) const {
-        return a == b;
+        return (a == b) || (hasAdditionalEqualities[a] &&
+                            hasAdditionalEqualities[b] &&
+                            equalitySet.count(std::pair<AlphabetIdx, AlphabetIdx>(a, b)));
     }
 };
 
+/**
+ * Takes additional equalities for alphabet elements, transforms them into index equalities
+ * @param [in] additionalEqualitiesLength
+ * @return  equalitySet as a set of all equality pairs where each pair contains the indexes of an additional equality
+ */
+template<class Element, class AlphabetIdx>
+EqualityDefinition<AlphabetIdx> makeEqualityDefinition(const EdlibAlignConfig<Element> config,
```

Comment:
I am somewhat confused with the implementation here,  you are taking both config and then additionalEqualities and their length! I would say you need either config or additionalEqualities and their length, not both at the same time, right?

ALso, below you have two for loops, why? Why not just have one for loop, merge them together?

Btw check this comments, I think this is the way to go: https://github.com/Martinsos/edlib/pull/154#discussion_r463247148 .